msprime/likelihood.py

```diff
@@ -0,0 +1,195 @@
+                                            tables.nodes[child].time)
+                if tables.nodes[child].time == 0:
+                    continue_downwards = False
+            else:
+                continue_downwards = False
+        continue_upwards = True
+        tmp_edge = edge
+        parent = tables.edges[edge].parent
+        child = tables.edges[edge].child
+        if parent not in edges_above:
+            continue_upwards = False
+        while continue_upwards:
+            count_up = 0
+            to_check = edges_above[parent]
+            for e in to_check:
+                if (tables.edges[e].left <= mutant_location and
+                        mutant_location < tables.edges[e].right):
+                    count_up += 1
+                    tmp_edge = e
+            count_down = 0
+            to_check = edges_below[parent]
+            for e in to_check:
+                if (tables.edges[e].left <= mutant_location and
+                        mutant_location < tables.edges[e].right):
+                    count_down += 1
+            if count_up == 1 and count_down == 1:
+                parent = tables.edges[tmp_edge].parent
+                child = tables.edges[tmp_edge].child
+                potential_branch_length += (tables.nodes[parent].time -
+                                            tables.nodes[child].time)
+                if parent not in edges_above:
+                    continue_upwards = False
+            else:
+                continue_upwards = False
+        ret += math.log(potential_branch_length / total_material)
+    return ret
+
+
+def log_arg_likelihood(arg, rho):
+    tables = arg.tables
+    number_of_lineages = arg.num_samples
+    number_of_links = number_of_lineages * arg.sequence_length
+    number_of_edges = len(tables.edges)
+    edge = 0
+    parent = tables.edges[edge].parent
+    time = 0
+    ret = 0
+    while edge < number_of_edges:
+        parent = tables.edges[edge].parent
+        rate = number_of_lineages * (number_of_lineages - 1) / 2 + number_of_links * rho
+        ret -= rate * (tables.nodes[parent].time - time)
+        time = tables.nodes[parent].time
+        child = tables.edges[edge].child
+        if tables.nodes[parent].flags == msprime.NODE_IS_RE_EVENT:
+            while tables.edges[edge].child == child:
+                if tables.edges[edge].parent != tables.edges[edge - 1].parent:
+                    gap = tables.edges[edge].left - tables.edges[edge - 1].right
+                edge += 1
+            number_of_links -= gap
+            number_of_lineages += 1
+            if gap == 0:
+                # we evaluate the density rather than the probability if there is no gap
+                gap = 1
+            if rho == 0:
+                ret -= float("inf")
+            else:
+                ret += math.log(rho * gap)
+        else:
+            segment_length_in_children = -tables.edges.left[edge]
+            while tables.edges[edge].child == child:
```

Comment:
more concise and idiomatic to say:
```python
while edge < num_edges and edges[edge].child == child:
    edge += 1
```
rather than having a break at the tail of the loop.

Reply:
For the sake of paranoia: the second condition won't get checked if the first one fails? I never remember which language does what in those cases.

Reply:
This is safe, ``and`` guarantees the order in which things are executed and short-circuits. C has the same behaviour.